src/_ert/forward_model_runner/client.py

```diff
@@ -21,112 +15,132 @@ class ClientConnectionError(Exception):
+
+    async def __aenter__(self) -> Self:
+        await self.connect()
+        return self
 
-    async def _send(self, msg: AnyStr) -> None:
-        for retry in range(self._max_retries + 1):
+    async def __aexit__(
+        self, exc_type: Any, exc_value: Any, exc_traceback: Any
+    ) -> None:
+        try:
+            await self.send(DISCONNECT_MSG)
+        except ClientConnectionError:
+            logger.error("No ack for dealer disconnection. Connection is down!")
+        finally:
+            self.socket.disconnect(self.url)
+            await self._term_receiver_task()
+            self.term()
+
+    def term(self) -> None:
+        self.socket.close()
+        self.context.term()
+
+    async def _term_receiver_task(self) -> None:
+        if self._receiver_task and not self._receiver_task.done():
+            self._receiver_task.cancel()
+            await asyncio.gather(self._receiver_task, return_exceptions=True)
+            self._receiver_task = None
+
+    async def connect(self) -> None:
+        self.socket.connect(self.url)
+        await self._term_receiver_task()
+        self._receiver_task = asyncio.create_task(self._receiver())
+        try:
+            await self.send(CONNECT_MSG, retries=1)
+        except ClientConnectionError:
+            await self._term_receiver_task()
+            self.term()
+            raise
+
+    async def process_message(self, msg: str) -> None:
+        raise NotImplementedError("Only monitor can receive messages!")
+
+    async def _receiver(self) -> None:
+        while True:
             try:
-                if self.websocket is None:
-                    self.websocket = await self.get_websocket()
-                await self.websocket.send(msg)
-                return
-            except ConnectionClosedOK as exception:
-                error_msg = (
-                    f"Connection closed received from the server {self.url}! "
-                    f" Exception from {type(exception)}: {exception!s}"
+                _, raw_msg = await self.socket.recv_multipart()
+                if raw_msg == ACK_MSG:
+                    self._ack_event.set()
+                else:
+                    await self.process_message(raw_msg.decode("utf-8"))
+            except zmq.ZMQError as exc:
+                logger.debug(
+                    f"{self.dealer_id} connection to evaluator went down, reconnecting: {exc}"
                 )
-                raise ClientConnectionClosedOK(error_msg) from exception
-            except (TimeoutError, InvalidHandshake, InvalidURI, OSError) as exception:
-                if retry == self._max_retries:
-                    error_msg = (
-                        f"Not able to establish the "
-                        f"websocket connection {self.url}! Max retries reached!"
-                        " Check for firewall issues."
-                        f" Exception from {type(exception)}: {exception!s}"
+                await asyncio.sleep(1)
```

Comment:
Ideally this should never happen - only if the socket is forcibly closed. So I thought to give it some slack to recover. Not sure though, maybe we don't need it. The idea is to yield, since if the socket is closed `await asyncio.sleep(1)` allows other tasks to be run.